src/pid-file/src/lib.rs

```diff
@@ -175,3 +183,48 @@ impl fmt::Display for Error {
 }
 
 impl std::error::Error for Error {}
+
+/// An extension of `PidFile` that also stores metadata about a process' ports
+///
+/// This is not meant to be used in production, it is to help orchestrate
+/// processes on local deployments.
+pub struct DevelopmentPidFile(PidFile);
+
+/// Contents of a `DevelopmentPidFile`
+pub struct DevelopmentPidContents {
+    /// Process PID
+    pub pid: u32,
+    /// Process port mappings
+    pub port_metadata: HashMap<String, i32>,
+}
+
+impl DevelopmentPidFile {
+    /// Attempts to open and lock the specified PID file, as well as write metadata
+    /// to the file.
+    ///
+    /// If the file is already locked by another process, it returns
+    /// `Error::AlreadyRunning`.
+    pub fn open<P: AsRef<Path>>(path: P, port_metadata: &str) -> Result<DevelopmentPidFile, Error> {
```

Comment:
There's something fishy about `open` taking a pre-serialized JSON string while `read` returns the deserialized JSON string. The violation of an abstraction, basically. I see why it is done like this, though, and I don't have an obvious way to fix it!

Reply:
Would it be crazy to deserialize the command line args and pass a `HashMap` to `open`, then have `open` re-serialize it? It fixes the abstraction but ends up doing some unnecessary work. I think it's worth it though because 1. It only happens once at start-up and 2. It's only for development environments. So the extra work isn't a huge deal.

Reply:
Oh, I missed the part where the orchestrator would write the port metadata file. In that case we wouldn't need to serialize twice. 

Though in terms of the lifetime for the port file, how would the orchestrator know when to delete the file? We could no longer just delete it when it goes out of scope. We could just never delete it and rely on new processes to delete old ones.

Reply:
> Though in terms of the lifetime for the port file, how would the orchestrator know when to delete the file? We could no longer just delete it when it goes out of scope. We could just never delete it and rely on new processes to delete old ones.

Wait, why can't you delete it when it goes out of scope? That only happens on an explicit `drop_service`, no?